website/blog/2021-11-14-Presto-Cluster-Memory-Tuning.md

```diff
@@ -0,0 +1,63 @@
+
+###Why use memory pool
+The System Pool is used for the memory used by the system. For example, when transferring data between machines, the buffer is maintained in the memory. This part of the memory is mounted under the name of the system. So, why do you need reserved memory? If there is no Reserved Pool, when there are a lot of queries and the memory space is almost used up, a query that consumes a lot of memory starts to run. But at this time, there is no memory space for the query to run, and the query has been in a suspended state, waiting for available memory. But after other small-memory queries are finished, new small-memory queries are added. Because the small memory query occupies a small amount of memory, it is easy to find the available memory. In this case, the large-memory query will hang until starved to death, usually gets timed-out.
+
+Therefore, in order to prevent this kind of starvation, a space must be reserved for a large memory query to run. The reserved space is equal to the maximum memory allowed by the query. Every second, Presto picks out a query with the largest memory usage and allows it to use the reserved pool to avoid that there is always no available memory for the query to run.
+
+###How big is the reserved pool?
+Presto has a couple of system configs on memory for the stability of the entire cluster:
+* `query.max-memory-per-node`: The maximum user memory allowed by a query on a single worker
+* `query.max-total-memory-per-node`: The maximum (user + system) memory allowed by a query on a single worker. If not set, by default it's 1/3 of heap size
+* `query.max-memory`: The maximum user memory allowed by a single query in the entire cluster
+* `query.max-total-memory`: The maximum (user + system) memory allowed by a single query in the entire cluster
+
+When these thresholds are exceeded, your query will be terminated with an insufficient memory error. Presto also has another memory related self-protection mechanism. When the general pool memory is used up, some operators will be set to a blocked state, allowing the operator to suspend for a while, thereby avoiding the entire system from crashing.
+
+##Memory Tuning
+
+Now for a cluster configured for concurrent queries, there are a few factors that should be considered when setting memory in a Presto Worker:
+
+* TotalMemory: Total amount of distributed memory a query may use
+* GeneralPool
+* ReservedPool
+* Other: Operating system, some other processes in the system, and some off-heap memory, some operations in calculations require off-heap memory, such as some decompression algorithms
+* `memory.heap-headroom-per-node`: There are some memory allocations that cannot be counted by the Presto framework, such as some memory allocations in the third party library
+* `query.max-memory`: total memory allowed for a single query
+* `query.initial-hash-partitions`: The number of hash partitions in distributed joins and aggregations
```

Comment:
`initial-hash-partitions` is deprecated because it implies the hash partitions fluctuate.  The more recent and common naming is `hash-partition-count`.

Reply:
yes thanks I was using the legacy config. will update the content